File: backend/shared/database/models/knowledge.py
"""Knowledge (RAG) models for document metadata and embedded chunks."""
from datetime import datetime, timezone
from enum import Enum
from typing import List, Optional, Dict, Any

from pydantic import BaseModel, Field


class KnowledgeSourceType(str, Enum):
    FILE = "file"
    URL = "url"
    TEXT = "text"


class KnowledgeStatus(str, Enum):
    PROCESSING = "processing"
    READY = "ready"
    ERROR = "error"


class KnowledgeDocument(BaseModel):
    """Metadata for a knowledge source document."""

    id: Optional[str] = None
    workspace_id: Optional[str] = None
    name: str
    source_type: KnowledgeSourceType
    storage_url: Optional[str] = None
    file_size: int = 0
    content_hash: str
    assigned_assistant_ids: List[str] = []
    status: KnowledgeStatus = KnowledgeStatus.PROCESSING
    error_message: Optional[str] = None
    token_count: int = 0
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    last_synced_at: Optional[datetime] = None

    def to_dict(self) -> Dict[str, Any]:
        data = self.model_dump()
        data["source_type"] = self.source_type.value
        data["status"] = self.status.value
        return data

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "KnowledgeDocument":
        if "_id" in data:
            data["id"] = str(data["_id"])
            del data["_id"]
        return cls(**data)


class KnowledgeChunk(BaseModel):
    """Embedded chunk tied to one knowledge document."""

    id: Optional[str] = None
    workspace_id: str
    document_id: str
    document_name: str
    assistant_ids: List[str] = []
    chunk_text: str
    embedding: List[float]
    token_count: int = 0

    def to_dict(self) -> Dict[str, Any]:
        data = self.model_dump()
        data.pop("id", None)
        return data

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "KnowledgeChunk":
        if "_id" in data:
            data["id"] = str(data["_id"])
            del data["_id"]
        return cls(**data)

File: frontend/src/types/knowledge.ts
export interface KnowledgeItem {
  id: string;
  name: string;
  type: "file" | "url" | "text";
  size: number;
  tokenCount: number;
  assignedAgents: string[];
  status: "processing" | "ready" | "error";
  createdAt: string;
  lastSyncedAt?: string;
  errorMessage?: string;
}

export interface KnowledgeDocumentApi {
  id: string;
  name: string;
  source_type: "file" | "url" | "text";
  file_size?: number;
  token_count?: number;
  assigned_assistant_ids?: string[];
  status: "processing" | "ready" | "error";
  created_at: string;
  last_synced_at?: string;
  error_message?: string;
}

export function toKnowledgeItem(doc: KnowledgeDocumentApi): KnowledgeItem {
  return {
    id: doc.id,
    name: doc.name,
    type: doc.source_type,
    size: doc.file_size || 0,
    tokenCount: doc.token_count || 0,
    assignedAgents: doc.assigned_assistant_ids || [],
    status: doc.status,
    createdAt: doc.created_at,
    lastSyncedAt: doc.last_synced_at,
    errorMessage: doc.error_message,
  };
}

export function formatFileSize(bytes: number): string {
  if (bytes === 0) return "0 B";
  const k = 1024;
  const sizes = ["B", "KB", "MB", "GB"];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(1)) + " " + sizes[i];
}
